src/tapdrvr.c

```diff
@@ -169,6 +159,18 @@ Routine Description:
         if (NDIS_STATUS_SUCCESS == status)
         {
             DEBUGP (("[TAP] Registered miniport successfully\n"));
+
+            //
+            // This lock protects the AdapterList.
+            //
+            GlobalData.Lock = NdisAllocateRWLock(GlobalData.NdisDriverHandle);
+            if (GlobalData.Lock == NULL)
```

Comment:
This doesn't look right. I think its be too late to create the lock here. NdisMRegisterMiniportDriver() will call AdapterCreate() callback (the InitializeHandlerEx callback) which will call tapAdapterContextAddToGlobalList() which needs the lock.

Reply:
Hang on...so NdisAllocateRWLock requires the handle returned by NdisMRegisterMiniportDriver, and NdisMRegisterMiniportDriver expects NdisAllocateRWLock to have been called already? That doesn't seem right, but I've seen worse...

Are you sure NdisMRegisterMiniportDriver calls AdapterCreate? I thought adapter creation was after DriverEntry returned. I can step through this code path tomorrow.

Reply:
See the docs on NdisMRegisterMiniportDriver for which callbacks are called from it. InitializeHandlerEx() will definitely get called and that is AdapterCreate() for us.

A way out is to implement SetOptions and create the lock there. May be that's how its supposed to be used in 6.20 when NULL handle is not an option (Win7).